pkgs/native_assets_cli/lib/src/api/build_output.dart

```diff
@@ -13,30 +13,99 @@ import '../model/metadata.dart' as model;
+  /// these timestamps.
   DateTime get timestamp;
+
+  @Deprecated('Use assets2')
   List<Asset> get assets;
+
+  /// The assets produced by this build.
+  ///
+  /// In dry runs, the assets for all [Architecture]s for the [OS] specified in
+  /// the dry run must be provided.
+  // TODO: Rename to `assets` after removing old one.
+  Iterable<Asset> get assets2;
+
+  @Deprecated('Use dependencies2')
   Dependencies get dependencies;
-  Metadata get metadata;
 
+  /// The files used by this build.
+  ///
+  /// If any of the files in [dependencies2] is modified after [timestamp], the
+  /// build will be re-run.
+  // TODO: Rename to `dependencies` after removing old one.
+  Iterable<Uri> get dependencies2;
+
+  /// Metadata can to be passed to `build.dart` invocations of dependent
+  /// packages.
+  // TODO(dacoharkes): Then we also need to make the accessors and setters
+  // in BuildConfig methods.
+  T getMetadata<T>(String key);
+
+  /// Create a build output.
+  ///
+  /// The [timestamp] indicates the time the build this output belongs to
+  /// started. If omitted, [timestamp] defaults to the time the build started.
+  /// The [timestamp] is rounded down to whole seconds, because
+  /// [File.lastModified] is rounded to whole seconds and caching logic compares
+  /// these timestamps.
+  ///
+  /// The [Asset]s produced by this build or dry run can be passed in [assets]
+  /// or [addAssets]. In dry runs, the assets for all [Architecture]s for the
+  /// [OS] specified in the dry run must be provided.
+  ///
+  /// The files used by this build must be passed in [dependencies2] or
+  /// [addDependencies]. If any of these files is modified after [timestamp],
+  /// the build will be re-run.
+  ///
+  /// Metadata can be passed to `build.dart` invocations of dependent packages
+  /// via [metadata2] or [addMetadata].
   factory BuildOutput({
     DateTime? timestamp,
-    List<Asset>? assets,
-    Dependencies? dependencies,
-    Metadata? metadata,
+    Iterable<Asset>? assets,
+    @Deprecated('Use addDependencies.') Dependencies? dependencies,
+    Iterable<Uri>? dependencies2,
+    @Deprecated('Use addMetadata.') Metadata? metadata,
+    Map<String, Object>? metadata2,
   }) =>
       model.BuildOutput(
         timestamp: timestamp,
-        assets: assets?.map((e) => e as model.Asset).toList(),
-        dependencies: dependencies as model.Dependencies?,
-        metadata: metadata as model.Metadata?,
+        assets: assets?.cast<model.Asset>().toList(),
+        dependencies: dependencies2 != null
+            ? model.Dependencies(dependencies2.toList())
+            : dependencies as model.Dependencies?,
+        metadata: metadata2 != null
+            ? model.Metadata(metadata2)
+            : metadata as model.Metadata?,
       );
 
+  /// Add [Asset]s produced by this build or dry run.
+  ///
+  /// In dry runs, the assets for all [Architecture]s for the [OS] specified in
+  /// the dry run must be provided.
```

Comment:
@mosuem had the suggestion that we could instead give `Asset` an `OS` and `Architecture` field instead of `Target`. That would make it consistent with the separate fields in `BuildConfig` instead of combined fields.
And then the `Architecture` field can be optional. It must not be provided for dry run but must be provided for wet run.